### ISSUE TITLE
Bug: `no-unused-vars` false positive if func def w/ same name appears twice

### ISSUE THREAD

**Issue by adamlui** [eslint/eslint#19215]
### Environment

Node version: v22.9.0
npm version: v10.9.1
Local ESLint version: v9.16.0 (Currently used)
Global ESLint version: Not found
Operating System: win32 10.0.19045

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
import js from '@eslint/js'
import globals from 'globals'
import stylisticJS from '@stylistic/eslint-plugin-js'
import json from '@eslint/json'
import markdown from '@eslint/markdown'
import regexp from 'eslint-plugin-regexp'
import yml from 'eslint-plugin-yml'

export default [
    {
        files: ['**/*.js', '**/*.mjs'],
        languageOptions: {
            ecmaVersion: 'latest', sourceType: 'script',
            globals: {
                ...globals.browser, ...globals.greasemonkey, ...globals.node, chatgpt: 'readonly', chrome: 'readonly',
                config: 'writable', CryptoJS: 'readonly', dom: 'readonly', GM_cookie: 'readonly', hljs: 'readonly',
                icons: 'writable', ipv4: 'readonly', marked: 'readonly', modals: 'writable',
                renderMathInElement: 'readonly', settings: 'writable', sidebarToggle: 'writable'
            }
        },
        plugins: { regexp, 'js-styles': stylisticJS },
        rules: {
            ...js.configs.recommended.rules, ...regexp.configs['flat/recommended'].rules,
            'indent': 'off', 'no-unexpected-multiline': 'off', 'key-spacing': 'off', // allow whitespace anywhere
            'js-styles/no-trailing-spaces': 'error', // ...except at ends of lines
            'js-styles/max-len': ['error', { 'code': 120, // limit lines to 120 chars except if containing...
                'ignoreComments': true, 'ignoreStrings': true, // ...trailing/own-line comments, quoted strings...
                'ignoreTemplateLiterals': true, 'ignoreRegExpLiterals': true }], // ...or template/regex literals
            'js-styles/no-extra-semi': 'error', // disallow unnecessary semicolons
            'quotes': ['error', 'single', { 'allowTemplateLiterals': true }], // enforce single quotes except backticks to avoid escaping quotes
            'comma-dangle': ['error', 'never'], // enforce no trailing commas in arrays or objects
            'no-async-promise-executor': 'off', // allow promise executor functions to be async (to accomodate await lines)
            'no-constant-condition': 'off', // allow constant conditions
            'no-empty': 'off', // allow empty blocks
            'no-inner-declarations': 'off', // allow function declarations anywhere
            'no-useless-escape': 'off', // allow all escape chars cause ESLint sucks at detecting truly useless ones
            'no-unused-vars': ['error', { 'caughtErrors': 'none' }] // allow unused named args in catch blocks
        }
    },
    { files: ['**/chatgpt.js'], languageOptions: { globals: { chatgpt: 'off' }}},
    { files: ['**/*.mjs'], languageOptions: { sourceType: 'module' }},
    { files: ['**/*.json'], ignores: ['**/package-lock.json'], language: 'json/json', ...json.configs.recommended },
    {
        files: ['**/*.md'], language: 'markdown/commonmark', plugins: { markdown },
        rules: {
            ...markdown.configs.recommended[0].rules,
            'markdown/heading-increment': 'off', // allow headings to skip levels
            'markdown/fenced-code-language': 'off', // allow code blocks w/ no language specified
            'markdown/no-missing-label-refs': 'off' // allow missing label references
        }
    },
    { files: ['**/*.yaml, **/*.yml'], ...yml.configs['flat/standard'][1] }
]
```
</details>

```js
    const modals = {
...
        about() {
            log.caller = 'modals.about()'
            log.debug('Showing About modal...')

            // Show modal
            const aboutModal = siteAlert('',
                `🏷️ ${app.msgs.about_version}: <span class="about-em">${app.version}</span>\n`
                    + '⚡ ' + ( app.msgs.about_poweredBy ) + ': '
                        + `<a href="${app.urls.chatgptJS}" target="_blank" rel="noopener">chatgpt.js</a>`
                            + ` v${app.chatgptJSver}\n`
                    + '📜 ' + ( app.msgs.about_sourceCode )
                        + `: <a href="${app.urls.gitHub}" target="_blank" rel="nopener">`
                            + app.urls.gitHub + '</a>',
                [ // buttons
                    function checkForUpdates() { checkForUpdates() },
                    function getSupport(){},
                    function rateUs() { modals.open('feedback') },
                    function moreAIextensions(){}
                ], '', 656 // modal width
            )

            // Add logo
            const aboutHeaderLogo = logos.amzgpt.create() ; aboutHeaderLogo.width = 420
            aboutHeaderLogo.style.cssText = `max-width: 98% ; margin: 15px ${
                env.browser.isMobile ? 'auto' : '15.5%' } 17px`
            aboutModal.insertBefore(aboutHeaderLogo, aboutModal.firstChild.nextSibling) // after close btn

            // Center text
            aboutModal.querySelector('h2').remove() // remove empty title h2
            aboutModal.querySelector('p').style.cssText = (
                'justify-self: center ; text-align: center ; overflow-wrap: anywhere ;'
              + `margin: ${ env.browser.isPortrait ? '6px 0 -16px' : '3px 0 -6px' }` )

            // Hack buttons
            aboutModal.querySelectorAll('button').forEach(btn => {
                btn.style.cssText = 'height: 52px ; min-width: 136px'

                // Replace link buttons w/ clones that don't dismiss modal
                if (/support|extensions/i.test(btn.textContent)) {
                    const btnClone = btn.cloneNode(true)
                    btn.parentNode.replaceChild(btnClone, btn) ; btn = btnClone
                    btn.onclick = () => modals.safeWinOpen(app.urls[
                        btn.textContent.includes(app.msgs.btnLabel_getSupport) ? 'support' : 'relatedExtensions' ])
                }

                // Prepend emoji + localize labels
                if (/updates/i.test(btn.textContent))
                    btn.textContent = `🚀 ${app.msgs.btnLabel_updateCheck}`
                else if (/support/i.test(btn.textContent))
                    btn.textContent = `🧠 ${app.msgs.btnLabel_getSupport}`
                else if (/rate/i.test(btn.textContent))
                    btn.textContent = `⭐ ${app.msgs.btnLabel_rateUs}`
                else if (/extensions/i.test(btn.textContent))
                    btn.textContent = `🤖 ${app.msgs.btnLabel_moreAIextensions}`

                // Hide Dismiss button
                else btn.style.display = 'none'
            })

            log.debug('Success! About Modal shown')

            return aboutModal
        },
...
    }
...
    function checkForUpdates() {
        log.caller = 'checkForUpdates()'
        log.debug(`currentVer = ${app.version}`)

        // Fetch latest meta
        log.debug('Fetching latest userscript metadata...')
        xhr({
            method: 'GET', url: app.urls.update + '?t=' + Date.now(),
            headers: { 'Cache-Control': 'no-cache' },
            onload: resp => {
                log.debug('Success! Response received')

                // Compare versions, alert if update found
                log.debug('Comparing versions...')
                app.latestVer = /@version +(.*)/.exec(resp.responseText)[1]
                for (let i = 0 ; i < 4 ; i++) { // loop thru subver's
                    const currentSubVer = parseInt(app.version.split('.')[i], 10) || 0,
                          latestSubVer = parseInt(app.latestVer.split('.')[i], 10) || 0
                    if (currentSubVer > latestSubVer) break // out of comparison since not outdated
                    else if (latestSubVer > currentSubVer) // if outdated
                        return modals.open('update', 'available')
                }

                // Alert to no update found, nav back to About
                modals.open('update', 'unavailable')
        }})
    }
...
```

### What did you expect to happen?

`checkForUpdates()` def to pass lint for being used in `modals.about()`

### What actually happened?

` error  'checkForUpdates' is defined but never used  no-unused-vars`

Bug occurs because ESLint thought the named function `checkForUpdates()` inside array passed as 2nd arg of `siteAlert()` (wrapper for [`chatgpt.alert()`](https://github.com/KudoAI/chatgpt.js/blob/v3.3.5/chatgpt.js#L63-L279)) in `modals.about()` used to create button labels/callbacks is the one being used inside its own def, so the actually called one when def'd later looked unused to it.

Before this commit renamed `updateCheck()` to `checkForUpdates()`:

https://github.com/adamlui/ai-web-extensions/commit/23b16b954c43d11b54877802c23d68b9e66eb361

...effects of bug were not present since there was no 2 same-named functions for ESLint to mistake one as unused so lint passes 

### Link to Minimal Reproducible Example

https://github.com/adamlui/ai-web-extensions/commit/23b16b954c43d11b54877802c23d68b9e66eb361

### Participation

- [x] I am willing to submit a pull request for this issue.
